clientv3/watch.go

```diff
@@ -41,3 +46,480 @@ type WatchResponse struct {
+				if _, ok := cancelSet[pbresp.WatchId]; ok {
+					break
+				}
+				cancelSet[pbresp.WatchId] = struct{}{}
+				cr := &pb.WatchRequest_CancelRequest{
+					CancelRequest: &pb.WatchCancelRequest{
+						WatchId: pbresp.WatchId,
+					},
+				}
+				req := &pb.WatchRequest{RequestUnion: cr}
+				wc.Send(req)
+			}
+		// watch client failed to recv; spawn another if possible
+		// TODO report watch client errors from errc?
+		case <-w.errc:
+			if wc, wcerr = w.newWatchClient(); wcerr != nil {
+				w.errc <- wcerr
+				return
+			}
+			curReqC = w.reqc
+			if pendingReq != nil {
+				failedReq = pendingReq
+			}
+			cancelSet = make(map[int64]struct{})
+		case <-w.stopc:
+			w.errc <- nil
+			return
+		}
+
+		// send failed; queue for retry
+		if failedReq != nil {
+			go func() {
+				select {
+				case w.reqc <- pendingReq:
+				case <-pendingReq.ctx.Done():
+				case <-w.donec:
+				}
+			}()
+			failedReq = nil
+			pendingReq = nil
+		}
+	}
+}
+
+// dispatchEvent sends a WatchResponse to the appropriate watcher stream
+func (w *watcher) dispatchEvent(pbresp *pb.WatchResponse) bool {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+	ws, ok := w.streams[pbresp.WatchId]
+	if ok {
+		wr := &WatchResponse{*pbresp.Header, pbresp.Events}
+		ws.recvc <- wr
+	}
+	return ok
+}
+
+// serveWatchClient forwards messages from the grpc stream to run()
+func (w *watcher) serveWatchClient(wc pb.Watch_WatchClient) {
+	for {
+		resp, err := wc.Recv()
+		if err != nil {
+			select {
+			case w.errc <- err:
+			case <-w.donec:
+			}
+			return
+		}
+		select {
+		case w.respc <- resp:
+		case <-w.donec:
+			return
+		}
+	}
+}
+
+// serveCtxs removes a watcher when its watch context cancels
+func (w *watcher) serveCtxs() {
+	for {
+		cases := []reflect.SelectCase{}
+		wss := []*watcherStream{}
+
+		// build list of ctx done channels
+		w.mu.RLock()
+		sc := reflect.SelectCase{Dir: reflect.SelectRecv}
+		for _, ws := range w.streams {
+			wss = append(wss, ws)
+			sc.Chan = reflect.ValueOf(ws.initReq.ctx.Done())
+			cases = append(cases, sc)
+		}
+		w.mu.RUnlock()
+
+		// add control channels
+		sc.Chan = reflect.ValueOf(w.ctxc)
+		cases = append(cases, sc)
+		sc.Chan = reflect.ValueOf(w.donec)
+		cases = append(cases, sc)
+
+		idx, _, _ := reflect.Select(cases)
+		switch {
+		// w.ctxc; refresh list
+		case idx == len(cases)-2:
+		// w.donec; shutdown
+		case idx == len(cases)-1:
+			return
+		// stream cancelled
+		default:
+			w.mu.Lock()
+			w.closeStream(wss[idx])
+			w.mu.Unlock()
+			// lazily send cancel message if events on missing id
+		}
+	}
+}
+
+// serveStream forwards watch responses from run() to the subscriber
+func (w *watcher) serveStream(ws *watcherStream) {
+	emptyWr := &WatchResponse{}
+	wrs := []*WatchResponse{}
+	for {
+		curWr := emptyWr
+		outc := ws.outc
+		if len(wrs) > 0 {
+			curWr = wrs[0]
+		} else {
+			outc = nil
+		}
+		select {
+		case outc <- *curWr:
+			newRev := wrs[0].Events[0].Kv.ModRevision
+			if newRev != ws.lastRev {
```

Comment:
newRev should be from the last event (bug here), but it doesn't solve the coalescing problem since in this case newRev = 2 since rev 3 hasn't happened yet